### PR TITLE
Support `FieldSpec_Dynamic` and skip indexing null values

### PR DESCRIPTION
**Pull request by oshadmi**
Related #3484
redisearch api allows to define an index field which is dynamic - a field which can have more than a single type, e.g., TAG and NUMERIC (such as used by [RedisGraph](https://github.com/RedisGraph/RedisGraph/blob/25d4743d934d765bd9484a7ad1294626bc9c9a74/src/index/index.c#L72))

This was broken in #2985 by adding a `union` [here](https://github.com/RediSearch/RediSearch/pull/2985/files#diff-9d5f1ee78b7f9d217087511b15f259495cf1c874b16c3995be3b418e054662deR12)

This PR restores the support for `FieldSpec_Dynamic` and in addition it skips null values from being indexed
(null values could have been indexed wrongly, e.g., as numeric 0)

### CODE REVIEW

File: src/document.c [hunk 1]
```diff
@@ -674,17 +679,18 @@ FIELD_PREPROCESSOR(geoPreprocessor) {
     }
     fdata->numeric = geohash;
   } else if (str_count > 1) {
+    fdata->isMulti = 1;
     arrayof(double) arr = array_new(double, str_count);
     for (size_t i = 0; i < str_count; ++i) {
       const char *cur_str = DocumentField_GetArrayValueCStr(field, &len, i);
       if ((parseGeo(cur_str, len, &lon, &lat) != REDISMODULE_OK) || ((geohash = calcGeoHash(lon, lat)) == INVALID_GEOHASH)) {
         array_free(arr);
+        fdata->arrNumeric = 0;
```
Comment (ashtul) [inline src/document.c:688]: = null;

File: src/document.c [hunk 2]
```diff
@@ -373,7 +375,7 @@ void AddDocumentCtx_Free(RSAddDocumentCtx *aCtx) {
         TagIndex_FreePreprocessedData(aCtx->fdatas[ii].tags);
         aCtx->fdatas[ii].tags = NULL;
       } else if (FIELD_IS(aCtx->fspecs + ii, INDEXFLD_T_GEO) && aCtx->fdatas[ii].isMulti &&
-                 aCtx->fdatas[ii].arrNumeric) {
+                 aCtx->fdatas[ii].arrNumeric && !FIELD_IS_NULL(aCtx, ii)) {
```
Comment (ashtul) [inline src/document.c:378]: is needed?

File: src/tag_index.h [hunk 3]
```diff
@@ -116,8 +117,10 @@ void TagIndex_Free(void *p);
 
 char *TagIndex_SepString(char sep, char **s, size_t *toklen);
 
-/* Preprocess a document tag field, returning a vector of all tags split from the content */
-char **TagIndex_Preprocess(char sep, TagFieldFlags flags, const DocumentField *data);
+/* Preprocess a document tag field, split the content in data into fdata `tags` array
+   Return 0 if there's no content in the field, 1 otherwise
```
Comment (ashtul) [inline src/tag_index.h:121]:
```suggestion
   Return 0 if the field contains a Null value, 1 otherwise
```

**Review by ashtul — APPROVED** [review]

File: src/document.c (hunk 2 above)
Comment (oshadmi) [reply src/document.c:378]: Needed for null geo values

**Review by rafie — APPROVED** [review]

**Pull request merged**